clash-prelude/tests/Clash/Tests/Fixed.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+module Clash.Tests.Fixed (tests) where
+
+import qualified Prelude as P
+import Test.Tasty
+import Test.Tasty.QuickCheck hiding (Fixed)
+
+import Clash.Prelude hiding (span)
+
+-- Saturate a number to be within certain bounds according to SaturationMode.
+saturate
+  :: RealFrac a
+  => a
+  -- minBound
+  -> a
+  -- maxBound
+  -> a
+  -- Span: the repetition interval of wrapping behaviour.
+  -> SaturationMode
+  -> a -> a
+saturate minB _ span SatWrap x =
+  let offs = x - minB
+      offsW = minB + (offs - span * fromIntegral (floor @_ @Integer
+                (offs / span)))
+  in offsW
+saturate minB maxB _ SatBound x
+  | x < minB  = minB
+  | x > maxB  = maxB
+  | otherwise = x
+saturate minB maxB _ SatZero x
+  | x < minB  = 0
+  | x > maxB  = 0
+  | otherwise = x
+saturate minB maxB _ SatSymmetric x
+  | x < minB  = if minB < 0 then (-maxB) else minB
+  | x > maxB  = maxB
+  | otherwise = x
+
+-- Saturate to bounds of type b.
+--
+-- It assumes the following values for the span (the range of numbers in the
+-- bounded type):
+--   - 2*(-minBound) when minBound < 0
+--   - floor maxBound + 1 otherwise
+-- This is correct for Unsigned, Signed and Fixed.
```

Comment:
This is confusing to me. Would span be "the number of numbers in the range"? If so, I would rename `span` to something like `nInhabitants` to make it a bit more obvious. Otherwise, if it is the "length" of the type's interval, wouldn't it simply be `abs (maxBound - minBound)`?

Reply:
No, it's the repetition interval of the type. For `Unsigned` we have:

```
minBound == 0
maxBound == 15
```
but wrapping behaves as modulo 16, not modulo 15.

Reply:
I'll rephrase it to use the phrase "repetition interval".

Reply:
Oh gotcha